django/db/models/deletion.py

```diff
@@ -408,7 +408,11 @@ def related_objects(self, related_model, related_fields, objs):
             [(f"{related_field.name}__in", objs) for related_field in related_fields],
             connector=query_utils.Q.OR,
         )
-        return related_model._base_manager.using(self.using).filter(predicate)
+        return (
+            related_model._base_manager.using(self.using)
+            .filter(predicate)
+            .order_by("pk")
+        )
```

Comment:
Doing the `Cast` when `lhs_field.db_type(conn) != rhs_field.db_type(conn)` has the side-effect that currently working requests are changed and will additionally do an unnecessary cast operation in their `ON`-clause.

I guess to avoid those things, we would need to know which types can be implicitly converted by which DB (a task that largely exceeds this patch IMHO).

-----

Why are we setting ordering here?

Without this, the test case `admin_utils.tests.NestedObjectsTests.test_siblings` fails.
See https://djangoci.com/job/pull-requests-focal/database=mysql,label=focal-pr,python=python3.11/19453/testReport/junit/admin_utils.tests/NestedObjectsTests/test_siblings/

```
Traceback (most recent call last):
  File "/home/jenkins/workspace/pull-requests-focal/database/mysql/label/focal-pr/python/python3.11/tests/admin_utils/tests.py", line 57, in test_siblings
    self._check([0, [1, 2]])
  File "/home/jenkins/workspace/pull-requests-focal/database/mysql/label/focal-pr/python/python3.11/tests/admin_utils/tests.py", line 38, in _check
    self.assertEqual(self.n.nested(lambda obj: obj.num), target)
AssertionError: Lists differ: [0, [2, 1]] != [0, [1, 2]]

First differing element 1:
[2, 1]
[1, 2]

- [0, [2, 1]]
+ [0, [1, 2]]
```

because we are adding a cast to the `on`-clause.
In the nested collector, prior to this patch, it used to fetch child elements by doing:
```
FROM `admin_utils_count` INNER JOIN `admin_utils_count` T2 ON (`admin_utils_count`.`parent_id` = T2.`id`) WHERE `admin_utils_count`.`parent_id` IN 1
```
And now we do:
```
FROM `admin_utils_count` INNER JOIN `admin_utils_count` T2 ON (`admin_utils_count`.`parent_id` = CAST(T2.`id` AS signed integer)) WHERE `admin_utils_count`.`parent_id` IN 1
```
with a cast to `signed integer`.

To me, this shows that the test is relying on an implicit ordering behaviour on JOINs.
Which is fixed by forcing the ordering on the PK here.
Even if this might have a slight performance impact.